core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala

```diff
@@ -1052,39 +1041,22 @@ class DAGScheduler(
               //       we registered these map outputs.
               mapOutputTracker.registerMapOutputs(
                 shuffleStage.shuffleDep.shuffleId,
-                shuffleStage.outputLocs.map(list => if (list.isEmpty) null else list.head),
+                shuffleStage.outputLocs.map(_.headOption.orNull),
                 changeEpoch = true)
 
               clearCacheLocs()
+
+              // Some tasks had failed; let's resubmit this shuffleStage
```

Comment:
Why move this outside the if-statement? It looks like it only holds when the if-statement is true, so less confusing I think to have it inside.